packages/gatsby/src/utils/api-runner-node.js

```diff
@@ -490,10 +490,16 @@ module.exports = async (api, args = {}, { pluginSource, activity } = {}) =>
     // call an action which will trigger the same API being called.
     // `onCreatePage` is the only example right now. In these cases, we should
     // avoid calling the originating plugin again.
-    const implementingPlugins = plugins.filter(
+    let implementingPlugins = plugins.filter(
       plugin => plugin.nodeAPIs.includes(api) && plugin.name !== pluginSource
     )
 
+    if (api === `sourceNodes` && args.pluginName) {
+      implementingPlugins = implementingPlugins.filter(
+        plugin => plugin.name === args.pluginName
+      )
+    }
```

Comment:
You can combine these...
```suggestion
    let implementingPlugins = plugins.filter(
      plugin =>
        plugin.nodeAPIs.includes(api) && 
        (
          api === `sourceNodes` && args.pluginName ?
            plugin.name === args.pluginName :
            plugin.name !== pluginSource 
        )
    )
```

or something

Reply:
Yeah I thought about it. I actually find it harder to reason about this way. But not emotionally attached or anything.